src/big/big_decimal.cr

```diff
@@ -0,0 +1,262 @@
+require "big_int"
+
+# A `BigDecimal` represents arbitrary precision decimals.
+#
+# It is internally represented by a pair of `BigInt` and `UInt64`: value and scale.
+# Value contains the actual value, and scale tells the decimal point place.
+# e.g. value=1234, scale=2 => 12.34
+#
+# The general idea and some of the arithmetic algorithms were adapted from
+# the MIT/APACHE -licensed https://github.com/akubera/bigdecimal-rs
+
+class InvalidBigDecimalException < Exception
+end
+
+struct BigDecimal
+  ZERO                       = BigInt.new(0)
+  TEN                        = BigInt.new(10)
```

Comment:
Out of curiosity: Why a distinguished constant for 10? Is it a convenience to later express scientific notation?

Edit: nevermind, I should have continued reading to see how you used it :).

Reply:
This simply to not instantiate TEN every time it is used (it is not primitive type, so its instantiation have a price). See https://github.com/crystal-lang/crystal/pull/4876#discussion_r135635824.